src/OpenTelemetry.Api/Context/BaggageContext.cs

```diff
@@ -0,0 +1,241 @@
+﻿// <copyright file="BaggageContext.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace OpenTelemetry.Context
+{
+    /// <summary>
+    /// Baggage context.
+    /// </summary>
+    public readonly struct BaggageContext : IEquatable<BaggageContext>
+    {
+        private static readonly RuntimeContextSlot<BaggageContext> RuntimeContextSlot = RuntimeContext.RegisterSlot<BaggageContext>("otel.baggage_context");
+        private static readonly Dictionary<string, string> EmptyBaggage = new Dictionary<string, string>();
+
+        private readonly Dictionary<string, string> baggage;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="BaggageContext"/> struct.
+        /// </summary>
+        /// <param name="baggage">Baggage key/value pairs.</param>
+        public BaggageContext(Dictionary<string, string> baggage)
+        {
+            this.baggage = baggage ?? EmptyBaggage;
+        }
+
+        /// <summary>
+        /// Gets or sets the current <see cref="BaggageContext"/>.
+        /// </summary>
+        public static BaggageContext Current
```

Comment:
How is the usage going to be? Will users have to set Current, and then it'll start flowing? A minimal example would help.

Reply:
Is it like:
```
BaggageContext.Current.SetBaggage("userid", 109)
```

Then in some processor
```
var id = BaggageContext.Current.GetBaggage("userid");
activity.SetAttribute("id", id);
```

Reply:
> How is the usage going to be? Will users have to set Current, and then it'll start flowing?

Baggage.Context.Current will return the default empty instance if nothing has been set. So you never have to initialize it, you can just start assigning things to it and all will be fine. Need to test this and make sure, but that's my intention :smile:

> A minimal example would help.

The spec has an optional context parameter on all the methods. If you don't specify it, it defaults to the "current" one. I translated that to some static helpers.

These things are equivalent:
```csharp
var baggageContext = BaggageContext.Current;

BaggageContext.GetBaggage();
BaggageContext.GetBaggage(baggageContext);
BaggageContext.Current.GetBaggage();

string value = BaggageContext.GetBaggage("name");
string value = BaggageContext.GetBaggage("name", baggageContext );
string value = BaggageContext.Current.GetBaggage("name");
```
So users can access with or without the Current.

CorrelationContext/Baggage in the spec is this immutable thing. Each time you add to it, or remove from it, you get a new context that automatically becomes the current one.

These test _should_ all pass:
```csharp
var context = BaggageContext.SetBaggage("name1, "value1");
Debug.Assert(context == BaggageContext.Current);

var context2 = BaggageContext.SetBaggage("name2, "value2");
Debug.Assert(context != BaggageContext.Current);
Debug.Assert(context2 == BaggageContext.Current);

Debug.Assert(BaggageContext.GetBaggage("name2") != null);
Debug.Assert(BaggageContext.GetBaggage("name2", context2) != null);

// Can use the optional context parameter to inspect the original context:
Debug.Assert(BaggageContext.GetBaggage("name2", context) == null);

BaggageContext.Current = context;
Debug.Assert(BaggageContext.GetBaggage("name2") == null);
```

> Then in some processor

The spec doesn't mention how CorrelationContext/Baggage is to be applied to spans. So I'm not going to go there on this PR. This PR will be about implementing what is in the spec. Then I'll do a follow-up PR or two with extensions. I'm thinking a BaggageProcessor to apply it to spans. Maybe something to fill in the gaps with #969.